File: java/core/src/main/java/network/aika/neurons/Synapse.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package network.aika.neurons;

import network.aika.Model;
import network.aika.activations.Activation;
import network.aika.Element;
import network.aika.bindingsignal.BindingSignal;
import network.aika.activations.Link;
import network.aika.type.Obj;
import network.aika.type.TypeRegistry;
import network.aika.type.relations.Relation;
import network.aika.typedefs.LinkDefinition;
import network.aika.typedefs.SynapseDefinition;
import network.aika.bindingsignal.BSType;
import network.aika.bindingsignal.Transition;
import network.aika.type.ObjImpl;
import network.aika.queue.QueueProvider;
import network.aika.queue.Timestamp;
import network.aika.misc.direction.Direction;
import network.aika.queue.Queue;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.DataInput;
import java.io.DataOutput;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Stream;

import static network.aika.queue.Timestamp.MAX;
import static network.aika.queue.Timestamp.MIN;
import static network.aika.neurons.RefType.SYNAPSE_IN;
import static network.aika.neurons.RefType.SYNAPSE_OUT;
import static network.aika.typedefs.LinkDefinition.SELF;

/**
 *
 * @author Lukas Molzberger
 */
public abstract class Synapse extends ObjImpl implements Element, QueueProvider {

    protected static final Logger log = LoggerFactory.getLogger(Synapse.class);

    protected int synapseId;
    protected NeuronReference input;
    protected NeuronReference output;

    protected boolean propagable;

    public Synapse(SynapseDefinition type) {
        super(type);
    }

    public Synapse(SynapseDefinition type, Neuron input, Neuron output) {
        this(type);
        link(input.getModel(), input, output);
    }

    @Override
    public Stream<Obj> followManyRelation(Relation rel) {
        if(rel == SynapseDefinition.LINK)
            return null;
        else
            throw new RuntimeException("Invalid Relation");
    }

    @Override
    public Obj followSingleRelation(Relation rel) {
        if(rel == SynapseDefinition.SELF)
            return this;
        else if(rel == SynapseDefinition.INPUT)
            return getInput();
        else if(rel == SynapseDefinition.OUTPUT)
            return getOutput();
        else
            throw new RuntimeException("Invalid Relation");
    }

    public int getSynapseId() {
        return synapseId;
    }

    public void setSynapseId(int synapseId) {
        this.synapseId = synapseId;
    }

    public Map<BSType, BindingSignal> transitionForward(Map<BSType, BindingSignal> inputBindingSignals) {
        Map<BSType, BindingSignal> outputTransitions = new HashMap<>();
        Transition[] transitions = ((SynapseDefinition)type).getTransition();
        for (Transition t : transitions) {
            BindingSignal bs = inputBindingSignals.get(t.from());
            if (bs != null) {
                outputTransitions.put(t.to(), bs);
            }
        }
        return outputTransitions;
    }

    public Synapse setPropagable(Model m, boolean propagable) {
        if(this.propagable != propagable)
            input.getNeuron(m).setModified();

        getInput(m).updatePropagable(output.getNeuron(m), propagable);
        this.propagable = propagable;

        return this;
    }

    public boolean isPropagable() {
        return propagable;
    }

    public final void setModified(Model m) {
        Neuron n = getStoredAt().getNeuron(m, this);
        if(n != null)
            n.setModified();
    }

    public void setInput(Neuron n) {
        this.input = new NeuronReference(n, SYNAPSE_IN);
    }

    public void setOutput(Neuron n) {
        this.output = new NeuronReference(n, SYNAPSE_OUT);
    }

    public Synapse link(Model m, Neuron input, Neuron output) {
        synapseId = output.getNewSynapseId();

        setInput(input);
        setOutput(output);

        link(m);

        return this;
    }

    public void link(Model m) {
        input.getNeuron(m).addOutputSynapse(this);
        output.getNeuron(m).addInputSynapse(this);
    }

    public void unlinkInput(Model m) {
        getInput(m).removeOutputSynapse(this);
    }

    public void unlinkOutput(Model m) {
        getOutput(m).removeInputSynapse(this);
    }

    public final Link createLink(Activation input, Activation output) {
        return createLink(input, transitionForward(input.getBindingSignals()), output);
    }

    public final Link createLink(Activation input, Map<BSType, BindingSignal> bindingSignals, Activation output) {
        if(output.hasConflictingBindingSignals(bindingSignals))
            return null;
        else if(output.hasNewBindingSignals(bindingSignals)) {
            output = output.branch(bindingSignals);

            output.linkIncoming(input);
        }

        return ((SynapseDefinition)getType())
                .getLink()
                .instantiate(this, input, output);
    }

    public final Direction getStoredAt() {
        return ((SynapseDefinition)getType()).getStoredAt();
    }

    public NeuronReference getInputRef() {
        return input;
    }

    public NeuronReference getOutputRef() {
        return output;
    }

    public Neuron getInput() {
        return getInput(
                output.getRawNeuron().getModel()
        );
    }

    public Neuron getInput(Model m) {
        if(input == null)
            return null;

        return input.getNeuron(m);
    }

    public Neuron getOutput() {
        return getOutput(
                input.getRawNeuron().getModel()
        );
    }

    public Neuron getOutput(Model m) {
        if (output == null)
            return null;

        return output.getNeuron(m);
    }

    @Override
    public void write(DataOutput out) throws IOException {
        out.writeUTF(getClass().getName());

        super.write(out);
        out.writeInt(synapseId);

        out.writeLong(input.getId());
        out.writeLong(output.getId());
    }

    public static Synapse read(DataInput in, TypeRegistry tr) throws IOException {
        short synTypeId = in.readShort();
        SynapseDefinition synapseDefinition = (SynapseDefinition) tr
                .getType(synTypeId);
        Synapse syn = synapseDefinition.instantiate();
        syn.readFields(in, tr);
        return syn;
    }

    @Override
    public void readFields(DataInput in, TypeRegistry tr) throws IOException {
        super.readFields(in, tr);
        synapseId = in.readInt();
        input = new NeuronReference(in.readLong(), SYNAPSE_IN);
        output = new NeuronReference(in.readLong(), SYNAPSE_OUT);
    }

    @Override
    public Timestamp getCreated() {
        return MIN;
    }

    @Override
    public Timestamp getFired() {
        return MAX;
    }

    public void delete(Model m) {
        if(log.isInfoEnabled())
            log.info("Delete synapse: " + this);

        if(input != null)
            getInput(m).removeOutputSynapse(this);
        if(output != null)
            getOutput(m).removeInputSynapse(this);
    }

    @Override
    public Queue getQueue() {
        return null; // TODO
    }

    @Override
    public String toString() {
        return type.getName() +
                " in:[" + (input != null ? input.toKeyString() : "X")  + "] " +
                " --> " +
                " out:[" + (output != null ? output.toKeyString() : "X") + "])";
    }

    @Override
    public String toKeyString() {
        return (input != null ? input.toKeyString() : "X") +
                " --> " +
               (output != null ? output.toKeyString() : "X");
    }
}

File: src/network/synapse.cpp
#include "network/synapse.h"
#include "network/types/synapse_type.h"
#include "network/types/neuron_type.h"
#include "network/direction.h" // Include the full NetworkDirection definition
#include "network/linker.h"
#include <iostream>
#include <stdexcept>
#include <limits>


Synapse::Synapse(SynapseType* type) : Object(type), synapseId(0), input(nullptr), output(nullptr), propagable(false), pairedSynapseInputSide(nullptr), pairedSynapseOutputSide(nullptr) {}

Synapse::Synapse(SynapseType* type, Neuron* input, Neuron* output)
    : Object(type), synapseId(output->getNewSynapseId()),
      input(new NeuronReference(input, RefType::SYNAPSE_IN)), 
      output(new NeuronReference(output, RefType::SYNAPSE_OUT)), 
      propagable(false), pairedSynapseInputSide(nullptr), pairedSynapseOutputSide(nullptr)
{
    initFields();
    link(input->getModel(), input, output);
}

RelatedObjectIterable* Synapse::followManyRelation(Relation* rel) const {
    throw std::runtime_error("Invalid Relation: " + rel->getRelationLabel());
}

Object* Synapse::followSingleRelation(const Relation* rel) const {
    if (rel->getRelationLabel() == "SELF") {
        return const_cast<Synapse*>(this);
    } else if (rel->getRelationLabel() == "INPUT") {
        return getInput();
    } else if (rel->getRelationLabel() == "OUTPUT") {
        return getOutput();
    } else if (rel->getRelationLabel() == "PAIR") {
        return pairedSynapseOutputSide;
    } else if (rel->getRelationLabel() == "PAIR_IN") {
        return pairedSynapseInputSide;
    } else if (rel->getRelationLabel() == "PAIR_OUT") {
        return pairedSynapseOutputSide;
    } else {
        throw std::runtime_error("Invalid Relation for Synapse: " + rel->getRelationLabel());
    }
}

int Synapse::getSynapseId() const {
    return synapseId;
}

void Synapse::setSynapseId(int synapseId) {
    this->synapseId = synapseId;
}

Synapse* Synapse::getPairedSynapseInputSide() const {
    return pairedSynapseInputSide;
}

void Synapse::setPairedSynapseInputSide(Synapse* pairedSynapseInputSide) {
    this->pairedSynapseInputSide = pairedSynapseInputSide;
}

Synapse* Synapse::getPairedSynapseOutputSide() const {
    return pairedSynapseOutputSide;
}

void Synapse::setPairedSynapseOutputSide(Synapse* pairedSynapseOutputSide) {
    this->pairedSynapseOutputSide = pairedSynapseOutputSide;
}

BindingSignal** Synapse::transitionForward(const BindingSignal** inputBindingSignals) {
    SynapseType* synapseType = static_cast<SynapseType*>(getType());
    
    // Get the output neuron to determine the number of binding signal slots
    Neuron* outputNeuron = getOutput();
    if (!outputNeuron) {
        return nullptr;
    }
    
    // Allocate output array using the neuron's method
    BindingSignal** outputBindingSignals = outputNeuron->createBindingSignalArray();
    
    // Get input and output neuron array sizes for bounds checking
    Neuron* inputNeuron = getInput();
    int inputBSSlots = inputNeuron->getNumberOfBSSlots();
    int outputBSSlots = outputNeuron->getNumberOfBSSlots();
    
    // Map each input binding signal forward
    for (int fromSlot = 0; fromSlot < inputBSSlots; fromSlot++) {
        if (inputBindingSignals[fromSlot] != nullptr) {
            int toSlot = synapseType->mapTransitionForward(fromSlot);
            if (toSlot >= 0 && toSlot < outputBSSlots) {
                outputBindingSignals[toSlot] = const_cast<BindingSignal*>(inputBindingSignals[fromSlot]);
            }
        }
    }
    
    return outputBindingSignals;
}

BindingSignal** Synapse::transitionBackward(const BindingSignal** outputBindingSignals) {
    SynapseType* synapseType = static_cast<SynapseType*>(getType());
    
    // Get the input neuron to determine the number of binding signal slots
    Neuron* inputNeuron = getInput();
    if (!inputNeuron) {
        return nullptr;
    }
    
    // Allocate input array using the neuron's method
    BindingSignal** inputBindingSignals = inputNeuron->createBindingSignalArray();
    
    // Get input and output neuron array sizes for bounds checking
    int inputBSSlots = inputNeuron->getNumberOfBSSlots();
    
    Neuron* outputNeuron = getOutput();
    int outputBSSlots = outputNeuron->getNumberOfBSSlots();
    
    // Map each output binding signal backward
    for (int toSlot = 0; toSlot < outputBSSlots; toSlot++) {
        if (outputBindingSignals[toSlot] != nullptr) {
            int fromSlot = synapseType->mapTransitionBackward(toSlot);
            if (fromSlot >= 0 && fromSlot < inputBSSlots) {
                inputBindingSignals[fromSlot] = const_cast<BindingSignal*>(outputBindingSignals[toSlot]);
            }
        }
    }
    
    return inputBindingSignals;
}


Synapse* Synapse::setPropagable(Model* m, bool propagable) {
    if (this->propagable != propagable) {
        input->getNeuron(m)->setModified();
    }
    
    getInput(m)->updatePropagable(output->getNeuron(m), propagable);
    this->propagable = propagable;
    
    return this;
}

bool Synapse::isPropagable() const {
    return propagable;
}

void Synapse::setModified(Model* m) {
    Neuron* n = getStoredAt()->getNeuron(m, this);
    if (n != nullptr) {
        n->setModified();
    }
}

void Synapse::setInput(Neuron* n) {
    input = new NeuronReference(n, RefType::SYNAPSE_IN);
}

void Synapse::setOutput(Neuron* n) {
    output = new NeuronReference(n, RefType::SYNAPSE_OUT);
}

Synapse* Synapse::link(Model* m, Neuron* input, Neuron* output) {
    synapseId = output->getNewSynapseId();
    
    setInput(input);
    setOutput(output);
    
    link(m);
    
    return this;
}

void Synapse::link(Model* m) {
    input->getNeuron(m)->addOutputSynapse(this);
    output->getNeuron(m)->addInputSynapse(this);
}

void Synapse::unlinkInput(Model* m) {
    getInput(m)->removeOutputSynapse(this);
}

void Synapse::unlinkOutput(Model* m) {
    getOutput(m)->removeInputSynapse(this);
}

Link* Synapse::createLink(Activation* input, Activation* output) {
    return static_cast<SynapseType*>(getType())
        ->getLinkType()
        ->instantiate(this, input, output);
}

bool Synapse::hasLink(Activation* input, Activation* output) const {
    if (!input || !output) {
        return false;
    }
    
    // Check if the input activation has an output link to the output activation via this synapse
    auto outputLinks = input->getOutputLinks(const_cast<Synapse*>(this));
    for (Link* link : outputLinks) {
        if (link && link->getOutput() == output && link->getSynapse() == this) {
            return true;
        }
    }
    
    return false;
}

NetworkDirection* Synapse::getStoredAt() const {
    return static_cast<SynapseType*>(getType())->getStoredAt();
}

NeuronReference* Synapse::getInputRef() const {
    return input;
}

NeuronReference* Synapse::getOutputRef() const {
    return output;
}

Neuron* Synapse::getInput() const {
    if (!output || !output->getRawNeuron()) {
        return nullptr;
    }
    return getInput(output->getRawNeuron()->getModel());
}

Neuron* Synapse::getInput(Model* m) const {
    if (!input) {
        return nullptr;
    }
    return input->getNeuron(m);
}

Neuron* Synapse::getOutput() const {
    if (!input || !input->getRawNeuron()) {
        return nullptr;
    }
    return getOutput(input->getRawNeuron()->getModel());
}

Neuron* Synapse::getOutput(Model* m) const {
    if (!output) {
        return nullptr;
    }
    return output->getNeuron(m);
}

long Synapse::getCreated() const {
    // Return minimum timestamp value
    return 0; // Using 0 as MIN for now
}

long Synapse::getFired() const {
    // Return maximum timestamp value
    return std::numeric_limits<long>::max(); // Using max long as MAX for now
}

void Synapse::deleteSynapse(Model* m) {
    std::cout << "Delete synapse: " << toString() << std::endl;
    
    if (input) {
        getInput(m)->removeOutputSynapse(this);
    }
    if (output) {
        getOutput(m)->removeInputSynapse(this);
    }
}

Queue* Synapse::getQueue() const {
    return nullptr; // Not implemented yet
}

std::string Synapse::toString() const {
    return getType()->getName() +
        " in:[" + (input ? input->toKeyString() : "X") + "] " +
        " --> " +
        " out:[" + (output ? output->toKeyString() : "X") + "])";
}

std::string Synapse::toKeyString() const {
    return (input ? input->toKeyString() : "X") +
        " --> " +
        (output ? output->toKeyString() : "X");
}